Allow combining pushmeta-based macros.

maleadt opened the pull request:
Combining pushmeta-based macros on a function currently fails in `pushmeta->findmeta` with the following error:

```
julia> Base.@pure @inline foo() = return nothing
ERROR: @inline foo() = begin 
            $(Expr(:line, 1, :none))
            return nothing
        end is not a function expression
 in eval(::Module, ::Any) at ./boot.jl:268
```

Expanding macros before calling `pushmeta` fixes this.

maleadt committed:
Allow combining pushmeta!-based macros.

Teach pushmeta! about encountering other macro calls,
and make it jump across all but the last argument.

maleadt commented:
Apparently this has a corner-case when calling `error` from a macro annotated with one of these `pushmeta!` based macros. For example:

```julia
macro annotate(ex)
    println("Before: ", ex)
    println("After: ", macroexpand(ex))
    esc(isa(ex, Expr) ? Base.pushmeta!(macroexpand(ex), :whatever) : ex)
end

macro foo()
    error(:foo)
end

@annotate function bar()
    @foo()
end
```

This errors, because of how `error` is expanded (taking the place of the entire function definition):
```
Before: function bar()
    @foo
end
After: $(Expr(:error, ErrorException("foo")))
ERROR: LoadError: $(Expr(:error, ErrorException("foo"))) is not a function expression
```

timholy commented:
Rather than expanding, could you check whether the expression is a `:macrocall`, and if so apply it to `args[2]` and then stick the macrocall back in front again?

maleadt commented:
That sounds like a good alternative indeed. I'll have to make it scan for the function block argument though, as some of these macro's could accept other arguments before the function definition (say, `@target ptx function...end`).

maleadt commented:
Rebased and changed to jump over macros as part of `pushmeta!` instead. It assumes the last argument to a function-annotating macro will be the actual function (ie. `@annotate ... function...end` and not `@annotate function...end otherargs`), which seemed reasonable to me.

Anybody against this change?

tkelman commented:
>  It assumes the last argument to a function-annotating macro will be the actual function

Can this be checked explicitly?

timholy commented:
@tkelman, `findmeta` already checks.

timholy commented:
LGTM

timholy commented:
Thanks @maleadt! A very nice improvement.
